### PR TITLE
Payload type multiplexing

### PR DESCRIPTION
**Pull request by murillo128**
@youennf PTAL

### CODE REVIEW

File: codec-agnostic-rtp-payload-format.md [hunk 1]
```diff
@@ -62,6 +62,39 @@ The spatial layers are sent in ascending order, with the same RTP timestamp, and
 If the application wants to encrypt content, the application will split media content into frames before encryption.
 Each frame will be encrypted independently, using {{SFrame}} for instance, and transmitted by the packetizer.
 
+Payload multiplexing
+=======================
+
+In order to reduce the number of payload type codes on the SDP exchange, a single payload type code for the generic packetization will be used for each media type. That requires to also carry the original payload type code for the negotiated media format that the frame belongs to.
+
+The associated payload type will be sent in a header extension. The payload of associated payload headerr extension element can be encoded using either the one-byte or two-byte header defined in [[RFC5285]].
```
Comment (youennf) [inline codec-agnostic-rtp-payload-format.md]: s/headerr/header/

File: codec-agnostic-rtp-payload-format.md [hunk 2]
```diff
@@ -62,6 +62,39 @@ The spatial layers are sent in ascending order, with the same RTP timestamp, and
 If the application wants to encrypt content, the application will split media content into frames before encryption.
 Each frame will be encrypted independently, using {{SFrame}} for instance, and transmitted by the packetizer.
 
+Payload multiplexing
+=======================
+
+In order to reduce the number of payload type codes on the SDP exchange, a single payload type code for the generic packetization will be used for each media type. That requires to also carry the original payload type code for the negotiated media format that the frame belongs to.
```
Comment (youennf) [inline codec-agnostic-rtp-payload-format.md]: s/will be used/can be used/

Comment (youennf) [reply codec-agnostic-rtp-payload-format.md]: s/That requires to also carry/That requires to identify the original payload type code for the negotiated media format that the frame belongs to directly from the RTP packets of the frame.

File: codec-agnostic-rtp-payload-format.md [hunk 3]
```diff
@@ -62,6 +62,39 @@ The spatial layers are sent in ascending order, with the same RTP timestamp, and
...
+```
+      0                   1                   2                   3
+      0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
+     +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+     |      ID       |     len=1     |S|     APT     |    0 (pad)    |
+     +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+```
+Figure 2: Sample Associated Payload Type Encoding Using the Two-Byte Header Format
+
+The APT value is the value is the payload type code of the associated format passed to the rtp generic packetizer before any transformation is applied. The S bit indicates if the media stream can be forwarded safely starting from this RTP packet. Typically, it will be set to 1 on the first rtp packet of an I-Frame video frame and in all rtp audio packets.
```
Comment (youennf) [inline codec-agnostic-rtp-payload-format.md:92]: s/is the value is/is the value of/

File: codec-agnostic-rtp-payload-format.md [hunk 4]
```diff
@@ -62,6 +62,39 @@ The spatial layers are sent in ascending order, with the same RTP timestamp, and
...
+
+Receivers MUST be ready to receive RTP packets with different associated payload types in the same way it would receive different payload type codes on the RTP packets.
```
Comment (youennf) [inline codec-agnostic-rtp-payload-format.md:94]: I am not against the S bit, but it seems like something that is not related to the scope of this header extension. For instance it does not really make sense for audio codecs.
As a first step, I would simplify this header extension by removing the S bit.
Or write the possibility to add a S bit in this header extension as a note.

File: codec-agnostic-rtp-payload-format.md (hunk 1 above)
Comment (murillo128) [reply codec-agnostic-rtp-payload-format.md]: fixed

File: codec-agnostic-rtp-payload-format.md (hunk 2 above)
Comment (murillo128) [reply codec-agnostic-rtp-payload-format.md]: changed

File: codec-agnostic-rtp-payload-format.md (hunk 4 above)
Comment (murillo128) [reply codec-agnostic-rtp-payload-format.md:94]: The PT is only 7 bits, so there is a free bit we could use for free. The S bit is needed for video when no SVC (i.e. no Dependency Descriptor)  is used. That would allow to reduce the overhead a bit, but I agree it can be a bit weird.

**Comment by murillo128** [comment]
I have just realized that we have to change the multiplexing from the header extension to the payload in order to make it work with RED as it is not able to recover the RTP header extensions.

**Comment by youennf** [comment]
That seems to be an argument against option 3.
It seems option 2 and option 4 do not have this issue.
Option 1 could be somehow updated by adding another PT dedicated to RED + generic PT.

**Comment by murillo128** [comment]
Option 1 does not have the issue either. To use option 3 we would need to
prepend the associated payload type to the payload content instead of
having it as a header extension.

**Comment by youennf** [comment]
I should probably take some time to read the RED spec. But anyway, here is my understanding:
With option 1, we define a PT1 for regular codec-specific packetization and PT2 for generic packetization, associated to PT1.
We also define PT3 for RED associated with PT1.
When receiver gets PT3 content, how can it know whether this is opaque content or not?
If we define a PT4 for RED associated with PT2, everything becomes crystal clear.

**Comment by murillo128** [comment]
Red just appends the rtp payload of previous packets in current one, but
the header extensions are lost. Same happens with ulpfec.

Red has a header block in which the original PT is sent to know what is the
original PT:

                      0 1 2 3 4 5 6 7
                     +-+-+-+-+-+-+-+-+
                     |0|   Block PT  |
                     +-+-+-+-+-+-+-+-+


What I don't like about option 1 is that we would be adding 3 more payload
types potentially to the SDP (generic packetization, rtx and red), given
the current usage of payload types, I am afraid that we would hit a wall
when trying to implement this.

I am still more prone to option 3 even if we have to prepend the APT on the
payload.

**Comment by youennf** [comment]
Even if we put the PT in the RED payload, the receiver will not know whether the RED payload is codec-specific content (without PT prepended) or generic content (with PT prepended).

I don't how feasible this is but, if we do not have this direct mapping (RED PT -> Generic PT), it seems that the same RTP header extension used for generic packets could be used for RED packets.

Also, if the content is encrypted and a RTP header extension containing the Opus TOC is used in case of using the generic PT, the same header extension would also be useful with RED+encrypted content.

As of RTC

> What I don't like about option 1 is that we would be adding 3 more payload types potentially to the SDP (generic packetization, rtx and red)

Agreed this is not great. Is RTX suffering from the same issue?
It seems that if a packet containing CVO header extension is lost, we would better retransmit the CVO information.

**Comment by youennf** [comment]
> Is RTX suffering from the same issue?

Looking at RTX implementation in libwebrtc, it seems that RTX would copy header extensions like CVO. It could potentially pass any other header extension.

Looking at RED implementation in libwebrtc, RED is considered as a codec. From what I see, the RED encoder outputs directly packets in the form defined in https://tools.ietf.org/html/rfc2198. The encryption is then applied.
Conversely, on decoding side, the transform would be applied before RED decoder is splitting the content.
AIUI, all options we are discussing would work for RED if we consider RED as an alternative codec like Opus, unlike RTX.

As a side note, the encryption might have a bad effect on how RED encoder computes free space to stuff its redundant data.
That might be a good thing to discuss in webrtc insertable streams repo.

**Comment by murillo128** [comment]
Not sure about RED implementation on libwebrtc, but from the RED RFC:

```
   An RTP packet containing redundant data shall have a standard RTP
   header, with payload type indicating redundancy.  The other fields of
   the RTP header relate to the primary data block of the redundant
   data.

   Following the RTP header are a number of additional headers, defined
   in the figure below, which specify the contents of each of the
   encodings carried by the packet.
```

That means that the header extensions only apply to the primary payolad of the RED packet. the rest of the secundary RED payloads only carry the previous encoding payload and any RTP header extension is lost.

We could say that if the generic packetization is using within red, all the secondary payloads must have the same associated payload type as the primary one (which is cover on the rtp header extension normally).


Regarding ulpfec, i was wrong, it can protect header extensions:

```
7.3.  FEC Header for FEC Packets
[...]
    in other words, the CSRC list, RTP extension, and
   padding of the media payload packets, if present, are "counted" as
   part of the payload
```

**Comment by murillo128** [comment]
should we merge this PR and open a new issue for RED handling?

**Review by youennf — APPROVED** [review]
Let's go then for now.
For RED, let's open a new issue. We need to decide when happens the transform in that case.

**Comment by youennf** [comment]
> That means that the header extensions only apply to the primary payolad of the RED packet. the rest of the secundary RED payloads only carry the previous encoding payload and any RTP header extension is lost.

That makes sense to me if we think RED is aggregating encrypted content.

**Comment by murillo128** [comment]
Yes, i agree
